sentry-android/src/main/java/com/joshdholtz/sentry/Sentry.java

```diff
@@ -965,4 +972,104 @@ public static JSONObject getStackTrace(Throwable t) throws JSONException {
         }
     }
 
+	private static Map<String, Object> ReadContexts(Context context) {
+		final Map<String, Object> contexts = new HashMap<>();
+		contexts.put("os", OsContext());
+		contexts.put("device", DeviceContext(context));
+		contexts.put("package", PackageContext(context));
+		return contexts;
+	}
+
+	/**
+	 * Read the device and build into a map.
+	 *
+	 * Not implemented:
+	 *   -  battery_level
+	 *   If the device has a battery this can be an integer defining the battery level (in
+	 *   the range 0-100). (Android requires registration of an intent to query the battery).
+	 *   - name
+	 *   The name of the device. This is typically a hostname.
+	 *
+	 *   @see https://docs.getsentry.com/hosted/clientdev/interfaces/#context-types
+     */
+	private static Map<String, String> DeviceContext(Context context) {
+		final Map<String, String> device = new HashMap<>();
+		try {
+			// The family of the device. This is normally the common part of model names across
+			// generations. For instance iPhone would be a reasonable family, so would be Samsung Galaxy.
+			device.put("family", Build.BRAND);
+
+			// The model name. This for instance can be Samsung Galaxy S3.
+			device.put("model", Build.PRODUCT);
+
+			// An internal hardware revision to identify the device exactly.
+			device.put("model_id", Build.MODEL);
+
+			final String architecture = System.getProperty("os.arch");
+			if (architecture != null) {
+				device.put("architecture", architecture);
+			}
+
+			final int orient = context.getResources().getConfiguration().orientation;
+			device.put("orientation", orient == Configuration.ORIENTATION_LANDSCAPE ?
+					"landscape" : "portrait");
+
+			// Read screen resolution in the format "800x600"
+			// Normalised to have wider side first.
+			final Object windowManager = context.getSystemService(Context.WINDOW_SERVICE);
+			if (windowManager != null && windowManager instanceof WindowManager) {
+				final DisplayMetrics metrics = new DisplayMetrics();
+				((WindowManager) windowManager).getDefaultDisplay().getMetrics(metrics);
+				device.put("screen_resolution",
+						String.format("%sx%s",
+								Math.max(metrics.widthPixels, metrics.heightPixels),
+								Math.min(metrics.widthPixels, metrics.heightPixels)));
+			}
+
+		} catch (Exception e) {
+			Log.e(TAG, "Error reading device context", e);
+		}
+		return device;
+	}
+
+	private static Map<String, String> OsContext() {
+		final Map<String, String> os = new HashMap<>();
+		try {
+			os.put("type", "os");
+			os.put("name", "Android");
+			os.put("version", Build.VERSION.RELEASE);
+			if (Build.VERSION.SDK_INT < 4) {
+				os.put("build", Build.VERSION.SDK);
+			} else {
+				os.put("build", Integer.toString(Build.VERSION.SDK_INT));
+			}
+			final String kernelVersion = System.getProperty("os.version");
+			if (kernelVersion != null) {
+				os.put("kernel_version", kernelVersion);
+			}
+
+		} catch (Exception e) {
+			Log.e(TAG, "Error reading OS context", e);
```

Comment:
The idea behind these three functions is to build up as much device data as possible, and if anything fails by throwing an exception, the `Map` can be returned with as many fields as possible.

I've logged any errors - I'm not sure we take any action - sending these errors to Sentry would be difficult.